### PR TITLE
[rllib] Deprecate policy optimizers

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Deprecate policy optimizers and remove their usage from the built in algorithms. They are still available for use by third party algorithms, but will be removed in a later release (after June).

Changes in this PR:
- Copy necessary files of rllib/optimizers into rllib/execution
- Remove uses of rllib/optimizers in rllib/agents/*
- Some fixes to APPO to properly update target network

### CODE REVIEW

File: doc/source/rllib-models.rst [hunk 1]
```diff
@@ -91,7 +91,7 @@ Instead of using the ``use_lstm: True`` option, it can be preferable use a custo
 Batch Normalization
 ~~~~~~~~~~~~~~~~~~~
 
-You can use ``tf.layers.batch_normalization(x, training=input_dict["is_training"])`` to add batch norm layers to your custom model: `code example <https://github.com/ray-project/ray/blob/master/rllib/examples/batch_norm_model.py>`__. RLlib will automatically run the update ops for the batch norm layers during optimization (see `tf_policy.py <https://github.com/ray-project/ray/blob/master/rllib/policy/tf_policy.py>`__ and `multi_gpu_impl.py <https://github.com/ray-project/ray/blob/master/rllib/optimizers/multi_gpu_impl.py>`__ for the exact handling of these updates).
+You can use ``tf.layers.batch_normalization(x, training=input_dict["is_training"])`` to add batch norm layers to your custom model: `code example <https://github.com/ray-project/ray/blob/master/rllib/examples/batch_norm_model.py>`__. RLlib will automatically run the update ops for the batch norm layers during optimization (see `tf_policy.py <https://github.com/ray-project/ray/blob/master/rllib/policy/tf_policy.py>`__ and `multi_gpu_impl.py <https://github.com/ray-project/ray/blob/master/rllib/execution/multi_gpu_impl.py>`__ for the exact handling of these updates).
```
Comment (sven1977) [inline doc/source/rllib-models.rst:94]: Can we switch this to tf.keras.layers.BatchNormalization instead?
We should not use tf.layers anymore.

File: rllib/agents/trainer_template.py [hunk 2]
```diff
@@ -1,38 +1,56 @@
...
-                  make_policy_optimizer=None,
-                  after_init=None,
-                  before_train_step=None,
-                  after_optimizer_step=None,
-                  after_train_result=None,
-                  collect_metrics_fn=None,
-                  before_evaluate_fn=None,
-                  mixins=None,
-                  execution_plan=None):
+def build_trainer(
```
Comment (sven1977) [inline rllib/agents/trainer_template.py:31]: Awesome. Can we add warnings if these are still passed in as not None?

File: rllib/BUILD [hunk 3]
```diff
@@ -2054,15 +2047,6 @@ py_test(
     args = ["--stop-timesteps=2000", "--run=QMIX"]
 )
 
-py_test(
```
Comment (sven1977) [inline rllib/BUILD:2057]: Why did this have to go?

File: rllib/tests/test_external_multi_agent_env.py [hunk 4]
```diff
@@ -64,32 +58,6 @@ def test_external_multi_agent_env_sample(self):
         batch = ev.sample()
         self.assertEqual(batch.count, 50)
 
-    def test_train_external_multi_agent_cartpole_many_policies(self):
```
Comment (sven1977) [inline rllib/tests/test_external_multi_agent_env.py:67]: Do we have a replacement for this test?

File: doc/source/rllib-models.rst (hunk 1 above)
Comment (sven1977) [inline doc/source/rllib-models.rst:94]: Can we switch this to tf.keras.layers.BatchNormalization instead?
We should not use tf.layers anymore.

File: rllib/BUILD (hunk 3 above)
Comment (sven1977) [inline rllib/BUILD:2057]: Why did this have to go?

File: rllib/agents/trainer_template.py [hunk 5]
```diff
@@ -1,38 +1,56 @@
...
+        name,
+        default_policy,
+        default_config=None,
+        validate_config=None,
+        get_initial_state=None,  # DEPRECATED
```
Comment (sven1977) [inline rllib/agents/trainer_template.py:36]: Awesome. Should we add warnings if these are still passed in as not None?

File: rllib/tests/test_external_multi_agent_env.py [hunk 6]
```diff
@@ -64,32 +58,6 @@ def test_external_multi_agent_env_sample(self):
...
-        n = 20
```
Comment (sven1977) [inline rllib/tests/test_external_multi_agent_env.py:68]: Do we have a replacement for this test?

**Review by sven1977 — CHANGES_REQUESTED** [review]

File: doc/source/rllib-models.rst (hunk 1 above)
Comment (ericl) [reply doc/source/rllib-models.rst:94]: Yeah, though when I looked in the example linked it was still using the old batch norm code. We can address in a follow up.

File: rllib/BUILD (hunk 3 above)
Comment (ericl) [reply rllib/BUILD:2057]: There were some issues with QMIX + APEX, and I just decided to remove it since we have no test cases or benchmarks showing this is a useful algorithm.

File: rllib/agents/trainer_template.py (hunk 5 above)
Comment (ericl) [reply rllib/agents/trainer_template.py:37]: Added

File: rllib/tests/test_external_multi_agent_env.py (hunk 6 above)
Comment (ericl) [reply rllib/tests/test_external_multi_agent_env.py:68]: I'm not sure this was that useful a test (it's just evaluating with 20 policies, and isn't actually testing external envs), so just decided to drop it.

File: doc/source/rllib-models.rst (hunk 1 above)
Comment (sven1977) [reply doc/source/rllib-models.rst:94]: Fair enough.

File: rllib/BUILD (hunk 3 above)
Comment (sven1977) [reply rllib/BUILD:2057]: Fair enough. Yeah, we should put some time into it to see, whether it actually works.

File: rllib/tests/test_external_multi_agent_env.py (hunk 6 above)
Comment (sven1977) [reply rllib/tests/test_external_multi_agent_env.py:68]: Ok.

**Review by sven1977 — APPROVED** [review]
Nice!!

**Comment by tsaoyu** [comment]
Found this PR while reading Eric's PhD thesis. Just curious why deprecate the `policy optimizers` abstraction as this PR only mentioned what have been done.
